README.adoc
image::stoat-logo.png[align="center"]

Stoat - Realtime Function Static Analysis
-----------------------------------------

Stoat is a tool to assert properties about the callgraph of programs/libraries.
The primary usage of this tool is to analyze programs which need to perform
hard realtime operations in a portion of a mixed codebase.
This is done by identifying all functions which can transitively be called from
some known root function which must be realtime.
If any unsafe function which could block for an arbitrary amount of time is
found in this transitive closure, then an error is emitted to indicate where the
improper behavior can be found and what backtrace is responsible for it getting
called.

Why?
~~~~

Maintaining a large codebase in C/C\++ can make it very difficult to know what
code ends up calling what other routines.
This is further complicated when there is some segregation within one codebase
which may not be at all clear in implementation.
This is further complicated by the opaqueness of some C++ techniques, such as
virtual overloading, operator overloading, multiple inheritance, and
implicit conversions.

Requires
~~~~~~~~

- LLVM 3.3+
- Clang 3.3+
- c++filt
- ruby
- graphviz gem (OPTIONAL) for callgraph renderings
- colorize gem (OPTIONAL) for colored output

How To Build
~~~~~~~~~~~~

[source,shell]
-----------------------------------------------------------
git clone https://github.com/fundamental/stoat && cd stoat
mkdir build && cd build
cmake .. && make && make test
make install #or just run in place
-----------------------------------------------------------

Features
~~~~~~~~

- Inline safety markers +++__attribute__((annotate("realtime")))+++
- Inline hazard markers +++__attribute__((annotate("non-realtime")))+++
- Out-of-line whitelists/blacklists
- Out-of-line suppression files
- Graph capabilities for safe or contradicting graph
- Support for dispatch tree analysis within librtosc

How To Use
~~~~~~~~~~

1. Annotate a root realtime function with +++__attribute__((annotate("realtime")))+++
2. Compile all files with CC=stoat-compile CXX=stoat-compile++ or just compile
   files with clang's -emit-llvm flag
3. Run the stoat on the resulting llvm IR files using the --recursive option
4. View the contradictions
5. Optionally redo the analysis using the -G option to graphically view the
   contradictions
6. Update your code, a whitelist, or suppression list
7. Enjoy the safer code

For a more explicit guide please see the tutorial at
http://log.fundamental-code.com/2014/08/15/stoat-tutorial-example

License
~~~~~~~

Stoat is available under the GPLv3 license


ChangeLog
2014-08-15 Mark McCurry
	- Version 1.0 Release
2015-03-12 Dominic Sacre
	- Add Colorized Output
2015-03-25 Dominic Sacre
	- Add Shorter Graph Representation
2015-04-24 Mark McCurry
	- Version 2.0 Release
	- Major Refactorization
	- Make Graphviz Optional
	- Update Default White/Blacklists
	- Add Basic Dump Output
	- Add Support For Try{}Catch{}
	- Add Support For Rtosc Callbacks
	- Add Logo Digitized By Harry van Haaren
	- Add Wildcard Support For Function Specs
	- Add stoat-compile/stoat-compile++ utility
	- Add Parallel Analysis
	- Add Namespace Support
	- Add Translation Unit Local Template Alias Support 
	- Add More Global Init Checks
	- Add Multiple Inheritance Vtable Support




test/run-test.rb
#!/usr/bin/ruby
#
# stoat - LLVM Based Static Analysis Tool
# Copyright (C) 2015 Mark McCurry
#
# This file is part of stoat.
#
# stoat is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# stoat is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with stoat.  If not, see <http://www.gnu.org/licenses/>.
#
# Argument:
# 1      - path to run.rb
# 2      - path to libfoo.so
# 3      - compiler to use
# 4      - whitelist to use
# 4..end - path to source files

if ARGV.length < 5
    $stderr.puts "Too Few Arguments"
    error 1
end

deductions  = ARGV[0]
llvm_lib    = ARGV[1]
compiler    = ARGV[2]
whitelist   = ARGV[3]
blacklist   = ARGV[4]
suppression = ARGV[5]


if(suppression != "nil")
    suppression = "-s #{suppression}"
else
    suppression = ""
end

args = ""
if(/\+\+/.match compiler)
    args = "-std=c++11" #Show off inlining
else
    args = "-std=c11"
end

source_files = ARGV[6...ARGV.length].join " "

`mkdir tmp-build-dir`
Dir.chdir 'tmp-build-dir'
`#{compiler} #{args} -emit-llvm -c #{source_files}`
puts `ruby #{deductions} --llvm-passes #{llvm_lib} -w #{whitelist} -b #{blacklist} #{suppression} -r .`
return_code = $?.to_i
Dir.chdir '..'
`rm -r tmp-build-dir`

puts "return code = '#{return_code}'"

if(return_code == 0)
    exit(0)
else
    exit(-1)
end

test/test-input.c
#define REALTIME __attribute__((annotate("realtime")))
#define NREALTIME __attribute__((annotate("!realtime")))

void undefined_function(void);

void bar(void)
{}

void REALTIME foo(int x)
{
    (void) x;
    bar();
    undefined_function();
}


void NREALTIME baz(void)
{}

int tall_number;

int REALTIME main()
{
    int barbar;
    foo(barbar);
    baz();
    return 0;
}


test/test-input3.c

//Function prototype (eg from a header)
void prototype_only(void);

void trivial_function(void)
{}

void calling_function(void)
{
    prototype_only();
    trivial_function();
}


lib/graph.rb
module GraphRender

    #Convert Number to 2 Nibble Hex Format
    #
    def self.to_2hex(x)
        s = x.to_s 16
        if(s.length == 1)
            "0" + s
        else
            s
        end
    end


    #Generate a random color for a graph edge
    #
    #@return [String] a color
    def self.random_color
        r = (rand * 255 * 255).to_i % ("CE".to_i 16)
        g = (rand * 255 * 255).to_i % ("DD".to_i 16)
        b = (rand * 255 * 255).to_i % ("EC".to_i 16)
        "#" + ([r,g,b].map{|x| to_2hex(x)}.join)
    end

    #TODO make this process faster
    def self.transitive_closure(callgraph, root, known = Set.new([]))
        if(known.include?(root))
            return Set.new([])
        end
        result = Set.new(callgraph.parents(root))
        result << root

        result2 = Set.new
        result.map do |x|
            if(!(known.include?(x) || result2.include?(x)))
                result2 = result2.union(
                    transitive_closure(callgraph, x, result2.union([root]).union(known)))
            end
        end
        result2.union result
    end


    Color_Contradicted = "red"
    Color_Unsafe       = "black"
    Color_Safe         = "green"
    Color_Deduced      = "yellow"


    def self.to_graph(deductions, callgraph, graphfile, minimal_graph, shorten)
        g = GraphViz::new("G", "rankdir"=>"LR")
        important_nodes = Set.new
        color_nodes     = Hash.new
        puts "Coloring Deductions..."
        deductions.each_with_index do |deduction,i|
            if(deduction.contradicted_p)
                color_nodes[i] ||= Color_Contradicted
                deduction.contradicted_by.each do |x|
                    color_nodes[x] = Color_Unsafe
                end
            elsif(deduction.realtime_p)
                color_nodes[i] ||= "green"
                callgraph.children(i).each do |x|
                    d = deductions[x]
                    if(d.realtime_p && !d.contradicted_p)
                        color_nodes[x] = "green"
                    end
                end
            end
        end

        #For a minimal graph *only* include the realtime functions that eventually
        #call a non-realtime or ambigious function
        if(minimal_graph)
            puts "Minimizing Graph..."
            important_nodes = Set.new

            color_nodes.each do |call,color|
                if(color == "black")
                    important_nodes = important_nodes.union(
                        transitive_closure(callgraph, call, important_nodes))
                end
            end
        else
            important_nodes = color_nodes.keys
        end



        puts "Generating Graph Nodes..."
        argh      = Hash.new
        node_list = Hash.new
        important_nodes.each do |key|
            if(color_nodes.has_key?(key))
                val = color_nodes[key]
                name = callgraph.node_list[key].demangled_short
                if(shorten)
                    name = shorten_name(name)
                    argh[key] = name
                end
                node_list[key] = g.add_nodes(name, "color"=> val,
                                             :shape=>:box)
            end
        end

        puts "Generating Graph Links..."
        dont_repeat = Hash.new
        important_nodes.each do |src|
            callgraph.children(src).each do |dest|
                if(important_nodes.include?(dest) && node_list.has_key?(src) &&
                  node_list.has_key?(dest))
                    if(deductions[dest].non_realtime_p)
                        #if(!dont_repeat.has_key?([argh[src],argh[dest]]))
                            g.add_edges(node_list[src], node_list[dest],
                                        "color"=>(random_color), "style"=>"bold")
                            dont_repeat[[argh[src],argh[dest]]] = true
                        #end
                    else
                        #if(!dont_repeat.has_key?([argh[src],argh[dest]]))
                            g.add_edges(node_list[src], node_list[dest],
                                        "color"=>(random_color), "style"=>"dashed")
                            dont_repeat[[argh[src],argh[dest]]] = true
                        #end
                    end
                end
            end
        end
        g.output(:png => graphfile)
    end
end


src/pass1.cpp
#include <llvm/Analysis/CallGraph.h>
#include <llvm/Support/raw_ostream.h>
#include <llvm/IR/Module.h>
#include <llvm/IR/Constants.h>
#include <llvm/IR/LLVMContext.h>
#include <llvm/Pass.h>
#include <cassert>
#include <cxxabi.h>
#include <sstream>

using namespace llvm;
using std::string;

namespace {

template<class T>
std::string stringFrom(T x)
{
    std::stringstream ss;
    ss << x;
    return ss.str();
}

template<class T>
std::string to_s(T x)
{
    return stringFrom(x);
}

std::string typeToString(llvm::Type *type)
{
    std::string data_str;
    llvm::raw_string_ostream ss(data_str);
    type->getScalarType()->getScalarType()->print(ss);
    string sss = ss.str();
    if(sss[1] == '"')
        return sss.substr(2,sss.length()-4);
    else
        return sss.substr(1,sss.length()-2);
}

struct DummyPass : public FunctionPass {
    static char ID; // Pass ID, replacement for typeid
    DummyPass() : FunctionPass(ID) {}

    void getAnalysisUsage(AnalysisUsage &AU) const override {
        AU.setPreservesAll();
    }

    std::string runOnCallInst(CallInst *call)
    {
        auto fn2  = call->getCalledFunction();
        auto it   = call->getCalledValue();
        auto it2  = dyn_cast<GlobalAlias>(it);
        auto load = dyn_cast<LoadInst>(it);
        string vtable_call_name = "";
        if(!fn2 && it2)
            fn2 = dyn_cast<Function>(it2->getOperand(0));
        if(!fn2 && load) { //This might be a vtable call
            auto getelm = dyn_cast<GetElementPtrInst>(load->getOperand(0));
            if(getelm && getelm->getNumOperands() == 2) { //this is at least a function pointer at an offset
                auto offset = dyn_cast<ConstantInt>(getelm->getOperand(1));
                auto load2  = dyn_cast<LoadInst>(getelm->getOperand(0));
                if(offset && load) {
                    size_t offset_val = offset->getZExtValue();
                    auto bitcast = dyn_cast<BitCastInst>(load2->getOperand(0));
                    if(bitcast)
                        vtable_call_name = typeToString(bitcast->getSrcTy()) + to_s(offset_val);
                }
            }
        }
        string s;
        if(fn2) {
            s = fn2->getName().str();
        } else if(!vtable_call_name.empty()) {
            s = vtable_call_name;
        } else {
            //fprintf(stderr, "right here...\n");
            //call->dump();
            //fprintf(stderr, "FN: ");
            //call->getCalledValue()->dump();
            //fprintf(stderr, "ARG0: ");
            //call->getOperand(0)->dump();
            //fprintf(stderr, "ARG1: ");
            //call->getOperand(1)->dump();
            return "";
        }
        if(s == "llvm.dbg.value" || s == "llvm.var.annotation"
                || s == "llvm.stackrestore" || s == "llvm.stacksave"
                || s == "llvm.va_start"
                || s == "llvm.va_end"
                || s == "llvm.lifetime.start"
                || s == "llvm.lifetime.end"
                || strstr(s.c_str(), "llvm.memcpy")
                || strstr(s.c_str(), "llvm.memmove")
                || strstr(s.c_str(), "llvm.memset")
                || strstr(s.c_str(), "llvm.pow")
                || strstr(s.c_str(), "llvm.sqrt")
                || strstr(s.c_str(), "llvm.") //TODO check for false captures
                || strstr(s.c_str(), "llvm.umul"))
            return "";
        return s;
    }

    bool runOnFunction(Function &Fn) override {
        std::vector<string> v;
        for(auto &bb:Fn) {
            for(auto &i:bb) {
                if(i.getOpcode() == Instruction::Call) {
                    auto s = runOnCallInst(dyn_cast<CallInst>(&i));
                    if(!s.empty())
                        v.push_back(s);
                } else if(i.getOpcode() == Instruction::Invoke) {
                    auto invoke = dyn_cast<InvokeInst>(&i);
                    if(invoke->getCalledFunction())
                        v.push_back(invoke->getCalledFunction()->getName().str());
                    //else
                    //    fprintf(stderr, "Oh no, not this again\n");
                }
            }
        }
        if(!v.empty()) {
            fprintf(stderr, "%s :\n", Fn.getName().str().c_str());
            for(auto x:v)
                fprintf(stderr, "    - %s\n", x.c_str());
        } else {
            fprintf(stderr, "%s :\n", Fn.getName().str().c_str());
            fprintf(stderr, "    - nil\n");
        }
        return false;
    }
};

struct DummyPass2 : public ModulePass {
    static char ID; // Pass ID, replacement for typeid
    DummyPass2() : ModulePass(ID) {}

    void getAnalysisUsage(AnalysisUsage &AU) const override {
        AU.setPreservesAll();
    }

    void handleNested(Value *v)
    {
        if(auto o = dyn_cast<Function>(v))
        {
            fprintf(stderr, "%s :\n", o->getName().str().c_str());
        } else if(auto s = dyn_cast<GlobalVariable>(v)) {
            fprintf(stderr, "    - %s\n", dyn_cast<ConstantDataArray>(s->getOperand(0))->getAsString().str().c_str());
        } else
            fprintf(stderr, "We have a dinosaur\n");


    }

    bool runOnModule(Module &m) override {
        auto g = m.getNamedGlobal("llvm.global.annotations");
        if(g) {
            auto gg = dyn_cast<ConstantArray>(g->getOperand(0));
            auto ggg = gg->getOperand(0);
            unsigned ops = ggg->getNumOperands();
            if(ops > 2)
                ops = 2;
            for(unsigned i=0; i<ops; ++i) {
                auto op = ggg->getOperand(i);
                if(auto o = dyn_cast<ConstantExpr>(op))
                    handleNested(o->getOperand(0));
            }
        }
        return false;
    }
};

//Extract Class Hierarcy
struct DummyPass3 : public FunctionPass {
    static char ID; // Pass ID, replacement for typeid
    DummyPass3() : FunctionPass(ID) {}

    void getAnalysisUsage(AnalysisUsage &AU) const override {
        AU.setPreservesAll();
    }

    string getMethodName(char *name)
    {
        char buffer[1024];
        memset(buffer, 0, 1024);
        if(!name)
            return "";
        char *pos = rindex(name, ':');
        char *btmp = buffer;
        if(!pos)
            return "";
        pos++;
        while(*pos && *pos != '(')
        {
            *btmp++ = *pos++;
        }
        return buffer;
    }

    bool isConstructorp(char *name)
    {
        //Isolate Class Name
        string cname = getMethodName(name);
        if(cname.empty())
            return false;
        if(strstr(name, (cname + "::" + cname + "(").c_str()))
            return true;
        return false;
    }

    string extractSuperClass(Instruction *this_ptr, BitCastInst *possible)
    {
        if(auto load = dyn_cast<LoadInst>(possible->getOperand(0))) {
            if(this_ptr == dyn_cast<Instruction>(load->getOperand(0))) {
                std::string data_str;
                llvm::raw_string_ostream ss(data_str);
                possible->getDestTy()->getScalarType()->getScalarType()->print(ss);
                if(ss.str().substr(0,6) == "%class")
                    return ss.str();
            }
        }
        return "";
    }

    void findSuperClasses(Function &Fn, string name)
    {
        //The first IR Element is an alloca for the this pointer
        //later there is a load of this %this.addr
        //if this load is cast into anything other than i8***, then that's a
        //super class

        std::vector<string> class_list;
        Instruction *this_ptr = NULL;
        for(auto &bb:Fn) {
            for(auto &I:bb) {
                if(!this_ptr && dyn_cast<AllocaInst>(&I))
                    this_ptr = &I;

                if(auto bitcast = dyn_cast<BitCastInst>(&I))
                {
                    auto super = extractSuperClass(this_ptr, bitcast);
                    if(!super.empty())
                    {
                        bool not_already_here = true;
                        for(auto x:class_list)
                            if(x == super)
                                not_already_here = false;
                        if(not_already_here)
                            class_list.push_back(super.substr(7, super.length()-8));
                    }
                }
            }
        }

        std::unique(class_list.begin(), class_list.end());
        if(!class_list.empty()) {
            fprintf(stderr, "%s:\n", name.c_str());
            for(auto x:class_list)
                fprintf(stderr, "    - %s\n", x.c_str());
        }


    }

    bool runOnFunction(Function &Fn) override {
        int status = 0;
        char *realname = abi::__cxa_demangle(Fn.getName().str().c_str(), 0, 0, &status);
        if(isConstructorp(realname)) {
            //fprintf(stderr, "I'm in a constructor '%s'\n", realname);
            findSuperClasses(Fn, getMethodName(realname));
        }
        return false;
    }
};

//Extract Vtable contents
struct DummyPass4 : public ModulePass {
    static char ID; // Pass ID, replacement for typeid
    DummyPass4() : ModulePass(ID) {}

    void getAnalysisUsage(AnalysisUsage &AU) const override {
        AU.setPreservesAll();
    }

    void handleVtable(const char *name, ConstantArray *v)
    {
        //Field 0 - NULL?
        //Field 1 - Typeinfo
        //Field 2 - Offset 0 virtual method
        unsigned ops = v->getNumOperands();
        if(ops >= 2)
            fprintf(stderr, "%s:\n", name);
        for(unsigned i=2; i<ops; ++i)
        {
            auto op       = v->getOperand(i);
            char *fname = NULL;
            if(!dyn_cast<ConstantPointerNull>(op)) {
                Function *function = NULL;
                auto alias    = dyn_cast<GlobalAlias>(op->getOperand(0));//dyn_cast<BitCastInst>(op);
                auto someth   = op->getOperand(0);
                if(alias)
                    function = dyn_cast<Function>(alias->getOperand(0));
                else
                    function = dyn_cast<Function>(someth);
                if(function)
                    fname = strdup(function->getName().str().c_str());
            }
            if(fname && strlen(fname) == 0)
                fname = NULL;
            char *tmp = fname;
            while(tmp && *tmp && isprint(*tmp))
                tmp++;
            if(tmp)
                *tmp = 0;
            fprintf(stderr, "    %d: %s\n", i-2, fname);
        }
    }

    bool runOnModule(Module &m) override {
        auto &gl = m.getGlobalList();
        for(auto &g:gl) {
            if(g.getName().str().substr(0,4) == "_ZTV") {
                const char *vtable_name = g.getName().str().c_str();
                int status = 0;
                char *realname = abi::__cxa_demangle(vtable_name, 0, 0, &status);
                if(!realname || strstr(realname, "__cxxabi"))
                    continue;
                char *tmp = realname;
                while(*tmp && isprint(*tmp))
                    tmp++;
                *tmp = 0;

                if(g.getNumOperands())
                    handleVtable(realname+11, dyn_cast<ConstantArray>(g.getOperand(0)));
            }
        }
        return false;
    }
};
}

char DummyPass::ID = 0;
char DummyPass2::ID = 0;
char DummyPass3::ID = 0;
char DummyPass4::ID = 0;
static RegisterPass<DummyPass> P1("dummy1", "do nothin");
static RegisterPass<DummyPass2> P2("dummy2", "do nothin");
static RegisterPass<DummyPass3> P3("dummy3", "do nothin");
static RegisterPass<DummyPass4> P4("dummy4", "do nothin");


run.rb
#!/usr/bin/ruby
require 'optparse'
require 'ostruct'
require 'set'
require 'yaml'
require 'pp'


#Modifies mapa s.t. it includes the mapb info
def merge_maps(mapa, mapb)
    if mapb
        mapb.each do |key, val|
            if(mapa.has_key?(key) && mapa[key].is_a?(Array))
                mapa[key].concat val
            else
                mapa[key] = val
            end
        end
    end
end


options = OpenStruct.new
options.whitelist = []
options.blacklist = []
options.unmangled = OpenStruct.new
options.unmangled.whitelist = []
options.unmangled.blacklist = []
options.root = "./"
options.lib  = "./src/libfoo.so"
options.dir  = []
options.recursive = false

OptionParser.new do |opts|
      opts.banner = "Usage: example.rb [options] FILES"

      opts.on("-w", "--whitelist FILE",
              "Define a Mangled Whitelist File") do |list|
          options.whitelist << list
      end

      opts.on("-b", "--blacklist FILE",
              "Define a Mangled Blacklist File") do |list|
          options.blacklist << list
      end

      opts.on("-r", "--recursive DIR",
              "Enable Recursive Search Mode") do |dir|
          options.recursive = true
          options.root = dir
      end

      opts.on("-l", "--llvm-passes LIB",
              "The Library Containing The Needed LLVM Passes") do |lib|
          options.lib = lib
      end
end.parse!

#White List/Black List Expansion
white_list = []
options.whitelist.each do |x|
    white_list.concat File.read(x).split
end

black_list = []
options.blacklist.each do |x|
    black_list.concat File.read(x).split
end

#If there are unmangled files, then mangle them and add them to the end of the
#mangled lists

#For each one of the bitcode files, run the pass based preprocessor
files = []
if(!ARGV.empty?)
    files.concat ARGV
end
if(options.recursive)
    rfiles = []
    rfiles.concat `find #{options.root} -type f | grep -e "\\.bc$"`.split
    rfiles.concat `find #{options.root} -type f | grep -e "\\.o$"`.split

    rfiles.each do |f|
        if(/LLVM/.match `file #{f}`)
            files << f
        end
    end
end

if(files.empty?)
    puts "There Are No Files To Process"
    exit 1
end

#p files
callgraph = Hash.new
function_props = Hash.new
class_high = Hash.new
vtable_information = Hash.new

files.each do |f|
    p "running #{f} file..."
    `opt -load #{options.lib} --extract-callgraph < #{f} > /dev/null 2> sfpv_output.txt`
    #puts File.read("sfpv_output.txt")
    ncallgraph = YAML.load_file "sfpv_output.txt"


    `opt -load #{options.lib} --extract-annotations < #{f} > /dev/null 2> sfpv_output.txt`
    #puts File.read("sfpv_output.txt")
    nfunc = YAML.load_file "sfpv_output.txt"

    `opt -load #{options.lib} --extract-class-hierarchy < #{f} > /dev/null 2> sfpv_output.txt`
    class_nhigh = YAML.load_file "sfpv_output.txt"

    `opt -load #{options.lib} --extract-vtables < #{f} > /dev/null 2> sfpv_output.txt`
    vtable_ninformation = YAML.load_file "sfpv_output.txt"

    merge_maps(callgraph, ncallgraph)
    merge_maps(function_props, nfunc)
    merge_maps(class_high, class_nhigh)
    merge_maps(vtable_information, vtable_ninformation)
end

symbol_list = Set.new
callgraph.each do |key,val|
    symbol_list << key
    val.each do |x|
        symbol_list << x
    end
    val = val.uniq
end

puts "Demangling #{symbol_list.length} Symbols..."
demangled_symbols = Hash.new
demangled_short   = Hash.new

def shorten_symbol(sym_)
    paren = 0
    finished = false

    sym = ""
    sym_.reverse.each_char do |x|
        if finished
            sym << x
        end
        if x == ')'
            paren = paren + 1
        end

        if(x == '(' && paren == 1)
            finished = 1
        elsif(x == '(')
            paren = paren - 1
        end
    end
    sym.reverse
end

def demangle(symbol_list, demangled_symbols, demangled_short)
    puts "Demangling #{symbol_list.length} Symbols..."
    f = File.new("tmp_thing.txt", "w")
    f.write(symbol_list.to_a.join("\n"))
    f.close
    demangled_list = `cat tmp_thing.txt | c++filt`.split("\n")
    puts "Resulting in #{demangled_list.length} Items..."
    tmp = 0
    symbol_list.each do |x|
        demangled_symbols[x] = demangled_list[tmp]
        tmp = tmp + 1
    end

    demangled_symbols.each do |key, value|
        m = shorten_symbol(value)
        if(m.empty?)
            demangled_short[key] = value
        else
            demangled_short[key] = m
        end
    end
end
demangle(symbol_list, demangled_symbols, demangled_short)

#pp demangled_short


reason_user_w  = "The Function Was Declared Realtime By A Whitelist"
reason_user_b  = "The Function Was Declared NonRealtime By A Blacklist"
reason_code_w  = "The Function Was Declared Realtime By A Code Annotation"
reason_code_b  = "The Function Was Declared NonRealtime By A Code Annotation"
reason_deduced = "The Function Was Deduced To Need To Be RealTime As It Was Called By A Realtime Function"
reason_none    = "Nom Deduction has occured"
reason_nocode  = "No Code Or Annotations, So The Function is Assumed Unsafe"

class DeductionChain
    attr_accessor :deduction_source, :reason, :realtime_p, :non_realtime_p, :has_body_p, :contradicted_p, :contradicted_by


    def initialize
        @deduction_source = nil
        @reason           = "No Deduction has occured"
        @realtime_p       = false
        @non_realtime_p   = false
        @has_body_p       = false
        @contradicted_p   = false
        @contradicted_by  = Set.new
    end
end

property_list = Hash.new
symbol_list.each do |x|
    property_list[x] = DeductionChain.new
end

puts "Doing Property List Stuff"

#Add information about finding source
callgraph2 = Hash.new
callgraph.each do |key,value|
    property_list[key].has_body_p = true
    if(!value.include? "nil")
        callgraph2[key] = value
    end
end
callgraph = callgraph2

#Add Anything That's On the function_props list
function_props.each do |key, value|
    if(property_list.include? key)
        if(value.include? 'realtime')
            property_list[key].realtime_p = true
            property_list[key].reason     = reason_code_w
        elsif(value.include? 'non-realtime')
            property_list[key].non_realtime_p = true
            property_list[key].reason         = reason_code_b
        end
    end
end

#Add WhiteList Information
property_list.each do |key, value|
    if(!value.realtime_p &&
       !value.non_realtime_p)
        if(white_list.include?(key) || white_list.include?(demangled_short[key]))
            value.realtime_p = true
            value.reason = reason_user_w
        end
    end
end

#Add BlackList Information
property_list.each do |key, value|
    if(!value.realtime_p &&
       !value.non_realtime_p)
        if(black_list.include?(key) || black_list.include?(demangled_short[key]))
            value.non_realtime_p = true
            value.reason = reason_user_b
        end
    end
end

#Add Any Known Virtual Calls
vtable_information.each do |key, value|
    value.each do |key2, value2|
        if(value2 != "(none)" && value2 != "__cxa_pure_virtual")
            new_key = "class.#{key}#{key2}"
            if(!callgraph.include? new_key)
                callgraph[new_key] = []
            end
            callgraph[new_key] << value2
            if(!property_list.include? new_key)
                property_list[new_key] = DeductionChain.new
                symbol_list << new_key
            end
            if(!property_list.include? value2)
                property_list[value2] = DeductionChain.new
                symbol_list << value2
            end
            property_list[new_key].has_body_p = true
        end
    end
end

#Add Calls Down the hierarchy [THIS IS BUGGED XXX]
class_high.each do |sub, supers|
    supers.each do |super_|
        50.times do |x|
            testing = "class.#{super_}#{x}"
            source = "class.#{sub}#{x}"
            if(symbol_list.include? testing)
                callgraph[testing] ||= []
                callgraph[testing] << source
                symbol_list << source
                if(!property_list.include? source)
                    property_list[source] = DeductionChain.new
                    property_list[source].has_body_p = true
                end
                property_list[testing].has_body_p = true
            end
        end
    end
end

#Add C++ABI Destructor/Constructor Chaining
symbol_list.each do |sym_|
    sym = sym_.to_s #TODO figure out how an integer got in this list
    if /D1Ev$/.match sym
        sym_mod = sym.gsub(/D1Ev$/, "D2Ev")
        if(symbol_list.include?(sym_mod) && !callgraph.include?(sym))
            callgraph[sym] = [sym_mod]
            property_list[sym].has_body_p = true
        end
    end
end

symbol_list.each do |sym_|
    sym = sym_.to_s
    if /C1E/.match sym
        sym_mod = sym.gsub(/C1E/, "C2E")
        if(symbol_list.include?(sym_mod) && !callgraph.include?(sym))
            callgraph[sym] = [sym_mod]
            property_list[sym].has_body_p = true
        end
    end
end



#Add no source stuff
property_list.each do |key, value|
    if(!value.has_body_p && !value.realtime_p && !value.non_realtime_p)
        value.non_realtime_p = true
        value.reason         = reason_nocode
    end
end

#Perform Deductions
do_stuff = true
while do_stuff
    do_stuff = false
    property_list.each do |key, value|
        if(!value.contradicted_p)
            if(value.realtime_p() && callgraph.include?(key))
                callgraph[key].each do |x|
                    if(property_list[x].non_realtime_p)
                        value.contradicted_p = true
                        value.contradicted_by << x
                        do_stuff = true
                    elsif(!property_list[x].realtime_p)
                        property_list[x].realtime_p = true
                        property_list[x].deduction_source = key
                        property_list[x].reason = reason_deduced
                        do_stuff = true
                    end
                end
            end
        end
    end
end



#Regenerate Demangled Symbols
demangled_symbols = Hash.new
demangled_short   = Hash.new
demangle(symbol_list, demangled_symbols, demangled_short)

error_count = 0
property_list.each do |key, value|
    if(value.contradicted_p)
        error_count = error_count+1
        pp demangled_symbols[key]
        pp value
        puts "##The Deduction Chain:"
        next_prop = value.deduction_source
        while(next_prop) do
            puts " - #{demangled_symbols[next_prop]} : #{property_list[next_prop].reason}"
            next_prop = property_list[next_prop].deduction_source
        end
        puts "##The Contradiction Reasons:"
        value.contradicted_by.each do |x|
            puts " - #{demangled_symbols[x]} : #{property_list[x].reason}"
        end
        puts "\n\n\n"
    end
end

puts "Total of #{error_count} error(s)"

require "graphviz"

def to_2hex(x)
    s = x.to_s 16
    if(s.length == 1)
        "0" + s
    else
        s
    end
end

def random_color
    r = (rand * 255 * 255).to_i % ("CE".to_i 16)
    g = (rand * 255 * 255).to_i % ("DD".to_i 16)
    b = (rand * 255 * 255).to_i % ("EC".to_i 16)
    "#" + ([r,g,b].map{|x| to_2hex(x)}.join)
end

g = GraphViz::new( "G" )
color_nodes = Hash.new
property_list.each do |key,val|
    if(val.contradicted_p)
        color_nodes[key] ||= "red"
        val.contradicted_by.each do |x|
            color_nodes[x] = "black"
        end
    elsif(val.realtime_p)
        color_nodes[key] ||= "green"
    end
end


node_list = Hash.new
color_nodes.each do |key,val|
    if(demangled_short.include?(key) && demangled_short[key] && demangled_short[key].length != 0)
        node_list[key] = g.add_node(demangled_short[key], "color"=> val)
    end
end

callgraph.each do |src, dests|
    dests.uniq.each do |dest|
        if(node_list.include?(src) && node_list.include?(dest))
            if(property_list[dest].non_realtime_p)
                g.add_edges(node_list[src], node_list[dest], "color"=>(random_color), "style"=>"bold")
            else
                g.add_edges(node_list[src], node_list[dest], "color"=>(random_color), "style"=>"dashed")
            end
        end
    end
end
g.output( :png => "sfpv_graphics.png" )
